FAKE NEWS DETECTION USING ML

The TF-IDF vectorizer converts collection of raw document into matrix of TF-IDF features using Passive Aggressive Classifier(Supervised ML).

TF=Term Frequency- Number of times a word aapears in document
IDF=Inverse Document Frequency- How significant a term is

In [2]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
df = pd.read_csv('I:/Jupyter Programs/news.csv')
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
X = df['text']
Y = df['label']

In [5]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=7)

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
#decides weightage of words

tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

#preprocessing of train data
tfidf_train = tfidf_vectorizer.fit_transform(X_train)

#preprocessing of test data
tfidf_test = tfidf_vectorizer.transform(X_test)


In [15]:
#classifier or algorithm to learn the model
from sklearn.feature_extraction.text import TfidfVectorizer

passive = PassiveAggressiveClassifier(max_iter=50)

passive.fit(tfidf_train, Y_train)
Y_pred = passive.predict(tfidf_test) #checking prediction

#accuracy of the model
score = accuracy_score(Y_test, Y_pred)
print(score)

#confusion matrix/kind of error calculation
confusion_matrix(Y_test, Y_pred)

0.9273875295974744


array([[588,  50],
       [ 42, 587]], dtype=int64)

RESULT- From matrix,
588-Fake news predicted
587-True news predicted